mailet/standard/src/main/java/org/apache/james/transport/mailets/Expires.java

```diff
@@ -0,0 +1,144 @@
+
+    private Optional<Duration> minAge = Optional.empty();
+    private Optional<Duration> maxAge = Optional.empty();
+    private Optional<Duration> defaultAge = Optional.empty();
+    
+    @Override
+    public void init() throws MessagingException {
+        minAge = parseDuration("minAge");
+        maxAge = parseDuration("maxAge");
+        defaultAge = parseDuration("defaultAge");
+
+        if (minAge.isEmpty() && maxAge.isEmpty() && defaultAge.isEmpty()) {
+            throw new MessagingException("Please configure at least one of minAge, maxAge, defaultAge");
+        }
+
+        if (isAfter(minAge, maxAge)) {
+            throw new MessagingException("minAge must be before maxAge");
+        }
+        if (isAfter(defaultAge, maxAge)) {
+            throw new MessagingException("defaultAge must be before maxAge");
+        }
+        if (isAfter(minAge, defaultAge)) {
+            throw new MessagingException("minAge must be before defaultAge");
+        }
+    }
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        ZonedDateTime now = timeSource.get();
+        MimeMessage message = mail.getMessage();
+        Optional<ZonedDateTime> expires = parseExpiresHeader(message);
+        if (expires.isPresent()) {
+            if (minAge.isPresent() && expires.get().isBefore(now.plus(minAge.get()))) {
+                setExpiresAfter(message, now, minAge.get());
+            } else
+            if (maxAge.isPresent() && expires.get().isAfter(now.plus(maxAge.get()))) {
+                setExpiresAfter(message, now, maxAge.get());
+            }
+        } else if (defaultAge.isPresent()) {
+            setExpiresAfter(message, now, defaultAge.get());
+        }
+    }
+
+    @Override
+    public String getMailetInfo() {
+        return "Expire Mailet";
+    }
+
+    private Optional<Duration> parseDuration(String param) {
+        String duration = getInitParameter(param);
+        if (duration == null) {
+            return Optional.empty();
+        } else {
+            return Optional.of(DurationParser.parse(duration, ChronoUnit.DAYS));
+        }
+    }
+    
+    private boolean isAfter(Optional<Duration> a, Optional<Duration> b) {
+        return a.isPresent() && b.isPresent() && a.get().compareTo(b.get()) > 0;
+    }
+    
+    private Optional<ZonedDateTime> parseExpiresHeader(MimeMessage message) {
+        try {
+            String[] expires = message.getHeader(EXPIRES);
+            if (expires == null || expires.length == 0) {
+                return Optional.empty();
+            } else {
+                DateTime dt = new DateTimeParser(new StringReader(expires[0])).parseAll();
+                return Optional.of(ZonedDateTime.of(
+                    dt.getYear(), dt.getMonth(), dt.getDay(),
+                    dt.getHour(), dt.getMinute(), dt.getSecond(), 0,
+                    ZoneOffset.ofHoursMinutes(dt.getTimeZone() / 100, dt.getTimeZone() % 100)));
```

Comment:
I assumed this was the right way of doing it, since MessageSearches uses DateTimeParser as well. In its defense, it nicely ignores any extra zone info some mail programs like to add, like (CEST) or [Berlin]. Plus it actually preserves the zone offset, giving us a proper DateTime. 
Of course, I can change it over to another parser. DateTimeFieldLenientImpl.PARSER seems to be somewhat strict for this purpose, would require additional input string sanitation before use. What about javax.mail.internet.MailDateFormat ?

Reply:
No I did open a PR in MIME4J since then to bring it under test. We should be good with it IMO
https://github.com/apache/james-mime4j/pull/70

Maybe the offset part of the story could be made less hacky though...

Reply:
Well, that is how the DateTimeParser handles the offset, just treats it as an integer value without further checks. At least ZoneOffset.ofHoursMinutes(hh, mm) will throw for values too far off the chart, like -9999. It will accept anything is hhmm range though, like -1234. I don't see how to make this less hacky than it is, except patching it at the source in DateTimeParser.

Reply:
Fair